### DISCUSSION TITLE
Feature to use cached previously synced version of Notebook in Read-Only mode

### DISCUSSION THREAD

**Discussion by jittygitty** [zim-desktop-wiki/zim-desktop-wiki#2348] [Ideas]
**Is your feature request related to a problem? Please describe.**
I use zimwiki on both linux and windows. I host my notebooks on remote drive, mapped. On windows if remote drive synology went to sleep there's quite a delay as it wakes up. A lot of times I simply want READ access to search notebook. For other reasons as well, redundancy etc, I would really like ability to of one button click to open Zim in READ-ONLY mode based on "cached" previously synced notebook. I really only need to connect to mapped drive to "Write/Edit" and otherwise prefer to use my local cached copy of Notebook in READ-ONLY.

**Describe the solution you'd like**
Feature whether built-in or via Plugin I'm not sure, where-by on every edit/write/save the data is also synced from "main" (Read/Write) working folder of notebook to the "CACHED" READ-ONLY version. So basically, there should be two copies of Notebook, a READ-ONY one with its own path, and the main READ-WRITE version, with a simple sync of a page (after a save) to sync from main read-write version of file to the Read-Only version location. And on opening zim, we can click to quickly select if to open the READ-ONLY mode or normal main read/write.

**Describe alternatives you've considered**
I thought of just using rsync, or freefilesync/lsyncd or other tools that can watch files/folders and run a sync command upon detecting changes, but it just seemed too messy and not ideal. At least it seems there should be a plugin that could facilitate what I'm trying to achieve.

**Additional context**
Or perhaps with some info/guidance, I may be able to hire a python dev to develop a plugin.

Note I think this could also be extended and be a good feature for maintaining a central (web) server-side version of zim notebook for editing, while everyone can have a local read-only but be able to write/save only while they have access to the central webserver hosting notebook.

Anyway, if there's a plugin that already does what I'm asking, would appreciate a link to it. thx!

**Comment by introt** [comment]
> For other reasons as well, redundancy etc, I would really like ability to of one button click to open Zim in READ-ONLY mode based on "cached" previously synced notebook.

You can use eg. https://syncthing.net/ to achieve the "previously synced notebook" part, and add another entry into `notebooks.list` for the `one button click` access. The read-only mode state is saved in `state.conf` ie. outside the notebook (as long as you have `shared=True` in `notebook.zim`), so after toggling it on for the copy it simply stays on.

Additionally, a custom tool can take the current page's path and open the other version via a simple string substitution of the path.

---

> Feature whether built-in or via Plugin I'm not sure, where-by on every edit/write/save the data is also synced from "main" (Read/Write) working folder of notebook to the "CACHED" READ-ONLY version.

This sounds like a filesystem level thing. One possible solution is to use https://rclone.org/ to mount your remote disk with caching enabled. See https://rclone.org/commands/rclone_mount/#vfs-file-caching for more details.

> Note I think this could also be extended and be a good feature for maintaining a central (web) server-side version of zim notebook for editing, while everyone can have a local read-only but be able to write/save only while they have access to the central webserver hosting notebook.

See the Version Control plugin. Version control systems (the `git` in GitHub) are made for the collaborative use case, allowing multiple people to edit their local copies in parallel and merging their work - including conflict resolution etc.

**Comment by jittygitty** [comment]
@introt Thanks, I appreciate the ideas. Though I could cobble together some solution myself, I got bigger issues to deal with for now. Maybe when I have free time I'll try and write a good description and hire someone to write a plugin.

But yea probably the plugin would trigger some rsync or simple copy from write-enabled zim-wiki (on mapped network drive) to the 'local disk' read-only cached mirror of the zim-wiki notebook every time there's an edit etc. I'm not sure if there's already a mechanism for doing this through zim though, so the plugin may have to wrap or communicate with some filesystem watcher tool as I think you said.

One could then further set filesystem level 'read-only' permissions on the read-only zim-wiki mirror notebook to make sure one doesn't accidently enable write on it from the zim-wiki interface. Then when you start zim-wiki you could choose between the read-only mirror notebook copy or the original write-enabled one.

So I guess my main question for now is if zim has some internal mechanism that plugins can hook into so that they are made aware each time there is an 'edit' or change to content in notebook. I'm guessing this might be easy/already-exist for adding/removing pages but maybe not for any 'edits' to existing pages/content?

Or is it blind to anything like that and we'd need tie in something like https://github.com/joh/when-changed or https://emcrisostomo.github.io/fswatch/?

**Comment by introt** [comment]
> One could then further set filesystem level 'read-only' permissions on the read-only zim-wiki mirror notebook to make sure one doesn't accidently enable write on it from the zim-wiki interface. Then when you start zim-wiki you could choose between the read-only mirror notebook copy or the original write-enabled one.

I think adding a `notebook.zim` configuration option to lock the notebook into read-only mode would be the clean way to approach this - making the files themselves read-only means whatever sync tool you're using needs to either modify the permissions or run as another user. You may also look into sandboxing, eg. using Firejail or the Flatpak packaged Zim and giving it read-only access to the directory holding the cached copy.

Re: "every time there's an edit etc." - I'm not sure whether that's going to work with multiple computers - if your Linux machine only syncs/caches when it opens or edits files, you're not going to get your Windows machine's edits cached (without opening those files while online). If you really do want to go that route, the Zeitgeist logger plugin[^1] hooks up to all major page opening and editing events, and can thus be used as a reference.

As for this issue, I'd like to move it over to the Discussions for better visibility - threads there get bumped up to the top of the page every time they're replied to - is that okay?

[^1]: https://github.com/zim-desktop-wiki/zim-desktop-wiki/blob/develop/zim/plugins/zeitgeist-logger.py

**Comment by jittygitty** [comment]
@introt Thanks, very good points, I notice I wasn't thinking to clearly on some things, I blame posting while sleep deprived. But yea you're right, file level read-only is messy and it can't be true read-only anyway otherwise you can't ever update that so-called read-only cached/mirror version of the notebook : D I also missed the problem with multiple users, and you're very right about that, solution should play nice with multiple devices (or users) using that same networked mapped notebook, even if I currently am the only user and using it from just a single device. (Like I'm already thinking of giving my wife a read-only samba/cifs network share user to read/sync from network mapped zimwiki and sync/mirror her a local copy.)

So maybe there still needs to be some rsync to sync from network-mapped notebook to the local cached 'read-only' one, that runs each time the cached/mirror notebook is opened, to get changes made by other devices?

The zeitgeist plugin you mentioned sounds interesting, I'll have to look into it!

For me the biggest pain is the slow spin-up and wait time for my synology to wake up and serve the networked mapped notebook, the wait gets really annoying. So this read-only local mirror/cache notebook would solve issue for me, but I think its best solution has wider usage with additional scenarios for multi-user/devices etc. 

(As a side-note I noticed many other things I do with Windows that should NOT require anything from that networked mapped drive, still slow everything down, its like windows has to scan all the drives (even mapped) ones when you don't need anything from it and so it slows down many tasks for no good reason.)

Oh and yea feel free to move this issue wherever you think is best, thanks

**Reply by introt** [reply]
> So maybe there still needs to be some rsync to sync from network-mapped notebook to the local cached 'read-only' one, that runs each time the cached/mirror notebook is opened, to get changes made by other devices?

If you're only using one device at a time, it might be best to sync at mount time - after that, your plugin should be able to take over.

> For me the biggest pain is the slow spin-up and wait time for my synology to wake up and serve the networked mapped notebook, the wait gets really annoying. So this read-only local mirror/cache notebook would solve issue for me

Given that you're willing to pay for this plugin to be developed, you might instead spend that money on a solid state drive to eliminate the waiting issue. SSDs are pretty cheap these days and likely outlast a constantly spun-up-and-down-again hard drive. Taking long(er than usual) to spin up may be a sign that the drive is about to die btw. Do remember that using a NAS doesn't eliminate the need for backups :)

>  but I think its best solution has wider usage with additional scenarios for multi-user/devices etc.

Zim's built around text files and directories which means that most scenarios have battle-tested solutions; here are some of my picks:

- Single user, single device & cloud: `rclone sync` to push local rw copy to cloud(=your Synology)
- Single user, multi-device (& cloud): Syncthing; great pair with Markor to edit your notebooks on Android
- Multi-user: A source control/version control system like `git`; `fossil`, `bzr` and `hg` are also supported by the Version Control plugin. There are nice graphical clients for both desktops and mobile devices to handle the eventual merge conflicts, and we may get some of that functionality in Zim via #1986 one day

**Reply by introt** [reply]
You can also mix and match multiple tools: use `Syncthing` to edit your notes across your devices and `rclone` to push a(n encrypted) copy to a 3rd party cloud provider for redundancy. On top - or rather, under - all that you can run a next-gen file system like `zfs` (or `btrfs`) on your Linux machine, and have a local backup server (eg. a Raspberry Pi - no need to go all fancy) pull periodical filesystem snapshots to itself for safekeeping against ransomware.

**Comment by jittygitty** [comment]
@introt Sorry I've been busy with some more urgent matters, thanks for the suggestions, I'm still thinking of the best way. Although I'd love to upgrade to SSD's it's too expensive for 8TB SSD :) 

Two 8TB SSD's will run me about $1,000 but even so, I guess if I wasn't hiring devs from overseas, just like the billion/trillion dollar companies even do, then the SSD's wouldn't seem like a bad deal.

On the other hand, I'd still want the plugin, since it's good to have a local uptodate cached copy, what if you got to grab the laptop and run, the sync gives you another free 'backup' so to speak.

As I get older, I'm trying to diminish reliance on zfs and complex setups at least for my personal home use where I'm the only sysadmin, in favor of a setup my wife can understand and manage. That's why I wanted a plugin managing most of this with maybe a simple "rsync" executable to do the sync when needed. I could sync to a raspberry pi or orange pi with nvme SSD but I'd rather stick with 8tb mirror on synology that's very simple to manage for anyone and not much upkeep if any.

I can't rely on sync on mount-time because I mount/map drive on windows boot-up and I tend to never reboot. I don't reboot, but Microsoft Windows sure thinks my work is not worth xxxx, and will readily kill/close everything and reboot my system anytime it sees fit. Just had a gazillion tabs open and many of them in incognito/private tabs (just to login to diff. gmails etc) that can't be recovered?

(It's funny that I may finally decide to rid my house of all 'windows', not because of anything else other than just this simple infuriating annoyance that my PC can decide to kick me out and reboot and I can't do anything about it (save a few postponements, it's so insulting/degrading jumping through such 'hoops'). We should have full control of when or if we want to update/reboot etc.)

You're right about ransomware though, and for that I was thinking either synology btrfs snapshots and/or other snapshot/revision scheme, and backup would be other synology at different location.

Thanks for the link about https://github.com/zim-desktop-wiki/zim-desktop-wiki/pull/178 that looks very interesting!

**Reply by introt** [reply]
Sorry, didn't realize you have an 8 terabyte Zim Wiki - or no free disk bay for an extra disk ;)

> what if you got to grab the laptop and run

You wouldn't necessarily have the changes made on the other machines, and couldn't edit your copy while out and about?

> I can't rely on sync on mount-time because I mount/map drive on windows boot-up and I tend to never reboot.

s/mount time/Zim start time/ or plain periodical - also Syncthing takes care of this for you

> Just had a gazillion tabs open and many of them in incognito/private tabs (just to login to diff. gmails etc) that can't be recovered?

https://addons.mozilla.org/en-US/firefox/addon/multi-account-containers/

> my PC can decide to kick me out and reboot and I can't do anything about it

There's a lot to be said about how the FSF communicates, but they do have a point when calling users of proprietary software "useds" - they're not the ones in control.

Snarks aside, I'd still be in favor of using the plain old version control plugin. Maybe set it up to commit and push changes on page save and pull when changing pages to stay as up to date as possible? Perhaps something similar to what some iOS users are doing with Obsidian using Shortcuts to sync their notes via Git when opening and closing the app: https://meganesulli.com/blog/sync-obsidian-vault-iphone-ipad/ (haven't used this personally, but seems to work fine for a few of my friends apart from the occasional merge conflict).

I'm totally with you on reducing complexity, but ultimately what you're doing (online-first offline-accessible knowledge base that isn't reachable via the internet) isn't really straight forward to implement in a dummy-proof way - especially for multiple users. I however wish you the best of luck with your endeavor!
